File: src/test/java/cube8540/oauth/authentication/oauth/security/provider/ClientCredentialsEndpointFilterTest.java
package cube8540.oauth.authentication.oauth.security.provider;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.HttpRequestMethodNotSupportedException;

import javax.servlet.FilterChain;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import static cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsProviderTestHelper.BASIC_AUTH_CLIENT_ID;
import static cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsProviderTestHelper.BASIC_AUTH_CLIENT_SECRET;
import static cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsProviderTestHelper.CLIENT_SECRET;
import static cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsProviderTestHelper.FILTER_PATH;
import static cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsProviderTestHelper.RAW_CLIENT_ID;
import static cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsProviderTestHelper.makeAuthentication;
import static cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsProviderTestHelper.makeAuthenticationManager;
import static cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsProviderTestHelper.makeBasicAuthenticationHttpServletRequest;
import static cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsProviderTestHelper.makeFilterChain;
import static cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsProviderTestHelper.makeGetHttpServletRequest;
import static cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsProviderTestHelper.makeHttpServletResponse;
import static cube8540.oauth.authentication.oauth.security.provider.ClientCredentialsProviderTestHelper.makeParameterHttpServletRequest;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@DisplayName("클라이언트 인증 엔드포인트 필터 테스트")
class ClientCredentialsEndpointFilterTest {

    @Test
    @DisplayName("ONLY_POST 속성이 true로 설정 되어 있을때 요청을 POST 이외로 시도 했을시")
    void whenConfigOnlyPostSetTrueAndRequestMethodOtherThanPost() {
        ClientCredentialsEndpointFilter filter = new ClientCredentialsEndpointFilter(FILTER_PATH);
        HttpServletRequest request = makeGetHttpServletRequest();
        HttpServletResponse response = makeHttpServletResponse();

        filter.setOnlyPost(true);

        assertThrows(HttpRequestMethodNotSupportedException.class, () -> filter.attemptAuthentication(request, response));
    }

    @Test
    @DisplayName("요청 정보에서 아이디를 찾을 수 없을때")
    void whenIdCannotFoundInRequest()  {
        ClientCredentialsEndpointFilter filter = new ClientCredentialsEndpointFilter(FILTER_PATH);
        HttpServletRequest request = makeParameterHttpServletRequest();
        HttpServletResponse response = makeHttpServletResponse();
        Authentication authentication = makeAuthentication();
        AuthenticationManager manager = makeAuthenticationManager(authentication);

        SecurityContextHolder.clearContext();
        filter.setAuthenticationManager(manager);
        when(request.getParameter("client_id")).thenReturn(null);

        assertThrows(BadCredentialsException.class, () -> filter.attemptAuthentication(request, response));
    }

    @Test
    @DisplayName("Basic Authentication 인증")
    void basicAuthentication() throws Exception {
        ClientCredentialsEndpointFilter filter = new ClientCredentialsEndpointFilter(FILTER_PATH);
        HttpServletRequest request = makeBasicAuthenticationHttpServletRequest();
        HttpServletResponse response = makeHttpServletResponse();
        Authentication authentication = makeAuthentication();
        AuthenticationManager manager = makeAuthenticationManager(authentication);
        ArgumentCaptor<Authentication> authenticationCaptor = ArgumentCaptor.forClass(Authentication.class);

        SecurityContextHolder.clearContext();
        filter.setAuthenticationManager(manager);

        filter.attemptAuthentication(request, response);
        verify(manager, times(1)).authenticate(authenticationCaptor.capture());
        assertEquals(BASIC_AUTH_CLIENT_ID, authenticationCaptor.getValue().getPrincipal());
        assertEquals(BASIC_AUTH_CLIENT_SECRET, authenticationCaptor.getValue().getCredentials());
    }

    @Test
    @DisplayName("매개 변수를 이용한 인증")
    void authenticationByParameter() throws Exception {
        ClientCredentialsEndpointFilter filter = new ClientCredentialsEndpointFilter(FILTER_PATH);
        HttpServletRequest request = makeParameterHttpServletRequest();
        HttpServletResponse response = makeHttpServletResponse();
        Authentication authentication = makeAuthentication();
        AuthenticationManager manager = makeAuthenticationManager(authentication);
        ArgumentCaptor<Authentication> authenticationCaptor = ArgumentCaptor.forClass(Authentication.class);

        SecurityContextHolder.clearContext();
        filter.setAuthenticationManager(manager);

        filter.attemptAuthentication(request, response);
        verify(manager, times(1)).authenticate(authenticationCaptor.capture());
        assertEquals(RAW_CLIENT_ID, authenticationCaptor.getValue().getPrincipal());
        assertEquals(CLIENT_SECRET, authenticationCaptor.getValue().getCredentials());
    }

    @Test
    @DisplayName("이미 인증이 완료된 요청 일때")
    void whenRequestHasAlreadyAuthenticated() throws Exception {
        ClientCredentialsEndpointFilter filter = new ClientCredentialsEndpointFilter(FILTER_PATH);
        HttpServletRequest request = makeParameterHttpServletRequest();
        HttpServletResponse response = makeHttpServletResponse();
        Authentication authentication = makeAuthentication();

        SecurityContextHolder.getContext().setAuthentication(authentication);

        Authentication result = filter.attemptAuthentication(request, response);
        assertEquals(authentication, result);
    }

    @Test
    @DisplayName("인증 성공")
    void successfulAuthentication() throws Exception {
        HttpServletRequest request = makeParameterHttpServletRequest();
        HttpServletResponse response = makeHttpServletResponse();
        FilterChain filterChain = makeFilterChain();
        Authentication authentication = makeAuthentication();
        ClientCredentialsEndpointFilter filter = new ClientCredentialsEndpointFilter(FILTER_PATH);

        filter.successfulAuthentication(request, response, filterChain, authentication);
        assertEquals(authentication, SecurityContextHolder.getContext().getAuthentication());
        verify(filterChain, times(1)).doFilter(request, response);
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/security/provider/ClientCredentialsEndpointFilterTest.kt
package cube8540.oauth.authentication.oauth.security.provider

import cube8540.oauth.authentication.oauth.security.OAuth2ClientDetailsService
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.web.HttpRequestMethodNotSupportedException
import java.util.*
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class ClientCredentialsEndpointFilterTest {

    private val filterPath = "/auth"
    private val manager: AuthenticationManager = mockk()

    private val filter = ClientCredentialsEndpointFilter(filterPath)

    init {
        filter.setAuthenticationManager(manager)
    }

    @Test
    fun `request authentication not using post method when only post set`() {
        val request: HttpServletRequest = mockk {
            every { method } returns HttpMethod.GET.toString()
        }
        val response: HttpServletResponse = mockk()

        filter.onlyPost = true

        val thrown = Assertions.catchThrowable { filter.attemptAuthentication(request, response) }
        Assertions.assertThat(thrown).isInstanceOf(HttpRequestMethodNotSupportedException::class.java)
    }

    @Test
    fun `request authentication when request client id is null`() {
        val request: HttpServletRequest = mockk(relaxed = true) {
            every { method } returns HttpMethod.POST.toString()
            every { getParameter("client_id") } returns null
        }
        val response: HttpServletResponse = mockk()

        val thrown = Assertions.catchThrowable { filter.attemptAuthentication(request, response) }
        Assertions.assertThat(thrown).isInstanceOf(BadCredentialsException::class.java)
    }

    @Test
    fun `request basic authentication`() {
        val basicAuthentication = "Basic ${Base64.getEncoder().encodeToString("clientId:clientPassword".toByteArray())}"
        val request: HttpServletRequest = mockk(relaxed = true) {
            every { method } returns HttpMethod.POST.toString()
            every { getHeader(HttpHeaders.AUTHORIZATION) } returns basicAuthentication
        }
        val response: HttpServletResponse = mockk()
        val authenticationResult: Authentication = mockk()

        val authenticationCaptor = slot<Authentication>()

        every { manager.authenticate(capture(authenticationCaptor)) } returns authenticationResult

        val result = filter.attemptAuthentication(request, response)
        Assertions.assertThat(authenticationCaptor.captured.principal).isEqualTo("clientId")
        Assertions.assertThat(authenticationCaptor.captured.credentials).isEqualTo("clientPassword")
        Assertions.assertThat(result).isEqualTo(authenticationResult)
    }

    @Test
    fun `authentication by parameter`() {
        val request: HttpServletRequest = mockk(relaxed = true) {
            every { method } returns HttpMethod.POST.toString()
            every { getParameter("client_id") } returns "clientId"
            every { getParameter("client_secret") } returns "clientPassword"
        }
        val response: HttpServletResponse = mockk()
        val authenticationResult: Authentication = mockk()

        val authenticationCaptor = slot<Authentication>()

        every { manager.authenticate(capture(authenticationCaptor)) } returns authenticationResult

        val result = filter.attemptAuthentication(request, response)
        Assertions.assertThat(authenticationCaptor.captured.principal).isEqualTo("clientId")
        Assertions.assertThat(authenticationCaptor.captured.credentials).isEqualTo("clientPassword")
        Assertions.assertThat(result).isEqualTo(authenticationResult)
    }

    @Test
    fun `request when already authenticated`() {
        val request: HttpServletRequest = mockk(relaxed = true) {
            every { method } returns HttpMethod.POST.toString()
            every { getParameter("client_id") } returns "clientId"
            every { getParameter("client_secret") } returns "clientPassword"
        }
        val response: HttpServletResponse = mockk(relaxed = true)
        val authentication: Authentication = mockk(relaxed = true) {
            every { isAuthenticated } returns true
        }

        SecurityContextHolder.getContext().authentication = authentication

        val result = filter.attemptAuthentication(request, response)
        Assertions.assertThat(result).isEqualTo(authentication)
    }

    @AfterEach
    fun clear() {
        SecurityContextHolder.clearContext()
    }
}